### ISSUE TITLE
Three questions

### ISSUE THREAD

**Issue by xiang1030** [denpamusic/lua-lcdproc#1]
hello，my english is poor!
Using your sample and readme code now, I find the following problems
1:README.md
"local lcd = LCDproc("localhost", 13666)" Need to be changed to "local lcd = LCDproc.new("localhost", 13666) "
2:README.md
if i use debug “local lcd = LCDproc.new("localhost", 13666, { timeout = 3, debug = true })”
There will be problems
err log:
>>> hello
<<< connect LCDproc 0.5.9 protocol 0.3 lcd wid 21 hgt 8 cellwid 6 cellhgt 8
>>> screen_add my_screen
<<< success
>>> widget_add my_screen one title
<<< success
>>> widget_set my_screen one {Title Text}
<<< success
>>> widget_add my_screen two string
<<< listen my_screen
>>> widget_add my_screen three string
<<< success
>>> widget_set my_screen three 1 3 {Second Line Text}
<<< success
>>> widget_add my_screen four string
<<< success
>>> widget_set my_screen four 1 4 {Third Line Text}
<<< success
<<< [poll] success
<<< [poll] ignore my_screen
<<< [poll] listen my_screen
lua: r.lua:17: attempt to index field 'two' (a nil value)
stack traceback:
        r.lua:17: in function 'fn'
        /usr/lib/lua/lcdproc.lua:27: in function 'do_events'
        /usr/lib/lua/lcdproc.lua:250: in function 'poll'
        r.lua:26: in main chunk
        [C]: ?
3: I try to use "Menu:add_action (id, text, result, hidden, prev)",but something's wrong
err log:
menu_add_item "" checkbox_id checkbox -value on -allow_gray true -text {checkbox_txt}
<<< success
Item:init-hello
>>> menu_add_item "" menu_leave_id menu -text {menu_leave_txt}
<<< success
table: 0x4ffc20 table: 0x5038e0 leave_y yes     false   nil
Item:init-hello
lua: /usr/lib/lua/lcdproc/menu.lua:66: attempt to call method 'create' (a nil value)
stack traceback:
        /usr/lib/lua/lcdproc/menu.lua:66: in function </usr/lib/lua/lcdproc/menu.lua:63>
        (tail call): ?
        /usr/lib/lua/lcdproc/menu.lua:724: in function 'add_action'
        demo2.lua:24: in main chunk
        [C]: ?
We look forward to your reply!QAQ

**Comment by denpamusic** [comment]
Hi,

Thank you for your feedback!

1. Both ways will work, since there is [__call metamethod in LCDproc](https://github.com/denpamusic/lua-lcdproc/blob/master/src/lcdproc.lua#L64). I prefer shorter `local lcd = LCDproc("localhost", 13666)` hovewer you may use constructor if you like.

2. This is probably oversight on my part, since server tries to listen to the screen before it's widgets are completely set. I'll fix this in the upcoming release.

3. I'll need to test this a bit further, since on first glance I've got no idea what's different about Action class and why it doesn't work for you, while Checkbox which inherits the same base class does🤔

Anyways expect an update in the next two to three days, that'll address your issues.

**Comment by xiang1030** [comment]
Hello
Thank you for your reply.
Here are some more questions.
1:if i use "shorter local lcd = LCDproc("localhost", 13666)", there will be problems here
err log:
lua: demo2.lua:7: attempt to call local 'LCDproc' (a table value)
stack traceback:
        demo2.lua:7: in main chunk
        [C]: ?
2:
This is the official demo,I hope you have a demo to implement 'action' and other functions for testing.
--[[
client_set name Parenttest
menu_add_item "" ask menu "Leave menus?" -is_hidden true
menu_add_item "ask" ask_yes action "Yes" -next _quit_
menu_add_item "ask" ask_no action "No" -next _close_

menu_add_item "" test menu "Test"
menu_add_item "test" test_action action "Action"
menu_add_item "test" test_checkbox checkbox "Checkbox"
menu_add_item "test" test_ring ring "Ring" -strings "one\ttwo\tthree"
menu_add_item "test" test_slider slider "Slider" -mintext "" -maxtext "" -value "50"
menu_add_item "test" test_numeric numeric "Numeric" -value "42"
menu_add_item "test" test_alpha alpha "Alpha" -value "abc"
menu_add_item "test" test_ip ip "IP" -v6 false -value "192.168.1.1"
menu_add_item "test" test_menu menu "Menu"
menu_add_item "test_menu" test_menu_action action "Submenu's action"
menu_add_item "test_menu" ask_yes1 action "Yes" -next _quit_
menu_add_item "test_menu" ask_no1 action "No" -next _close_

menu_set_item "" test -prev "ask"

menu_set_item "test" test_action -next "test_checkbox"
menu_set_item "test" test_checkbox -next "test_ring" -prev "test_action"
menu_set_item "test" test_ring -next "test_slider" -prev "test_checkbox"
menu_set_item "test" test_slider -next "test_numeric" -prev "test_ring"
menu_set_item "test" test_numeric -next "test_alpha" -prev "test_slider"
menu_set_item "test" test_alpha -next "test_ip" -prev "test_numeric"
menu_set_item "test" test_ip -next "test_menu" -prev "test_alpha"
menu_set_item "test" test_menu_action -next "_close_"
--]]

I tried to use your code to create official demo:
my code demo:
--[[
local LCDproc = require "lcdproc"
require "socket"

function Sleep(n)
   socket.select(nil, nil, n)
end
local lcd = LCDproc.new("localhost", 13666, { timeout = 3, debug = true })
lcd:set_name("Simple_Demo")

-- create clock screen
local screen = lcd:add_screen("clock_screen")
screen:add_title("one", "Simple Clock")
screen:add_string("time", 1, 2, os.date("%H:%M:%S"))
screen:add_string("wcx", 1, 3, 'wcx')
screen:add_number("zgp", 1, 1)

-- create demo menu
local menu_demo = lcd:add_menu()
menu_demo:create()
menu_demo:add_checkbox('checkbox_id', 'checkbox_txt', 'on', true, false)

menu_leave = menu_demo:add_menu('menu_leave_id', 'menu_leave_txt', false)
-- Menu:add_action (id, text, result, hidden, prev)
-- But there will be problems here, and if you don't do anything about this menu, you can't control it after clicking enter
-- menu_leave:add_action('leave_demo', 'menu_leave_demo', 'none', false)
-- menu_leave:add_action('leave_y', 'yes', 'close', false)
-- menu_leave:add_action('leave_n', 'no', 'quit', false)

menu_leave1 = menu_leave:add_menu('menu_leave_id1', 'menu_leave_txt1', false)
-- Menu:add_action (id, text, result, hidden, prev)
-- But there will be problems here, and if you don't do anything about this menu, you can't control it after clicking enter
-- menu_leave1:add_action('leave1_demo', 'menu_leave1_demo', 'none', false)
-- menu_leave1:add_action('leave1_y', 'yes', 'close', false)
-- menu_leave1:add_action('leave1_n', 'no', 'quit', false)

menu_leave:add_checkbox('checkbox_id1', 'checkbox_txt1', 'on', true, false)
menu_leave:add_ring('ring_id', 'ring_txt', '1', {'wcx', 'wcx1', 'wcx2'}, false)
menu_leave:add_slider('slider_id', 'slider_txt', 10, {0, 100}, {'empty', 'full'}, 1, false)
menu_leave:add_numeric('numeric_id', 'numeric_txt', 10, {0, 100}, false)
menu_leave:add_alpha('alpha_id', 'alpha_txt', 'a', nil, {0, 10}, {'upper', 'lower', 'digit'}, 'abcdegfASDF./!@', false)
menu_leave:add_ip('ip_ip', 'ip_txt', '192.168.36.1', false, false)
menu_leave:add_checkbox('checkbox_id2', 'checkbox_txt2', 'on', true, false)

-- create demo_2 menu
local menu_demo_2 = lcd:add_menu()
menu_demo_2:add_checkbox('checkbox_id22', 'checkbox_txt22', 'off', true, false)

local active = false

function listen_do()
    active = true
    screen.widgets.one:set_text('zgp...')
end
function ignore_do()
    active = false
end

-- toggle active variable if screen is currenly being shown or ignored
-- lcd:on_listen(function () active = true screen.widgets.one:set_text('zgp...') end)
-- lcd:on_ignore(function () active = false end)
lcd:on_listen(listen_do)
lcd:on_ignore(ignore_do)

while true do
  -- poll LCDproc server once per second
  lcd:poll()

  if active then
    Sleep(1)
    -- update time only when screen is active
    screen.widgets.one:set_text('wcx...')
    screen.widgets.time:set_text(os.date("%H:%M:%S"))
  end
end

-- close connection to LCDproc server
lcd:close()
--]]

3:If you have time, please make a menu demo for us to debug.

**Comment by denpamusic** [comment]
I've just released [version 1.0.9](https://github.com/denpamusic/lua-lcdproc/releases/tag/v1.0.9) which should hopefuly address your issues.

As far as your example is concerned, you **must not** use `Sleep(1)` here, as it'll break event handling.
```lua
while true do
  -- poll LCDproc server once per second
  lcd:poll()

  if active then
    Sleep(1)
    -- update time only when screen is active
    screen.widgets.one:set_text('wcx...')
    screen.widgets.time:set_text(os.date("%H:%M:%S"))
  end
end
```
Having `lcd:poll()` in your loop is already introducing 1 second delay, if you need more than that please use [`LCDproc.every(interval, fn)`](https://lua-lcdproc.denpa.pro/modules/lcdproc.html#LCDproc.every) helper.

```lua
while true do
  -- poll LCDproc server once per second
  lcd:poll()

  if active then
    -- update data every 10 seconds
    LCDproc.every("10s", function ()
       screen.widgets.one:set_text('wcx...')
       screen.widgets.time:set_text(os.date("%H:%M:%S"))
    end)
  end
end
```

You also **must not** call `Menu:create()` directly, `add_menu()` will do it for you
```lua
-- create demo menu
local menu_demo = lcd:add_menu()
menu_demo:create()
menu_demo:add_checkbox('checkbox_id', 'checkbox_txt', 'on', true, false)
```
should instead be
```lua
-- create demo menu
local menu_demo = lcd:add_menu()
menu_demo:add_checkbox('checkbox_id', 'checkbox_txt', 'on', true, false)
```

**Comment by xiang1030** [comment]
Thank you very much for your reply, which helped solve many of my problems！
